### DISCUSSION TITLE
Mathematical function/ Quadratic Bezier curve

### DISCUSSION THREAD

**Discussion by ExFallenSkyTear** [Rem0o/FanControl.Releases#718] [Ideas]
I think it would be nice to have a fan curve where you can enter a mathematical function to control the fan,
I had this idea because I wanted to control my fans with a EaseInOut like type of function, which is basically a quadratic bezier, but setting it up manually is a hassle.
![Immagine 2022-04-30 184449](https://user-images.githubusercontent.com/33699967/166114638-6d975f1c-ac72-4676-a9c9-89d07651a4cc.png)

**Comment by Simone100190** [comment]
The curve itself is beautiful! BUT, do you think it is really a good curve?
I am using this software from a couple of days and i am trying to understand what is a good curve. 
The "ease in" (low temp section) i think is kinda good: as temperature rises, the fans are going faster and faster, in an "accelerated" way.
But looking at "ease out" (high temp section) i see the fans "decelerating" in a way, so it seems that as the temperature rises more and more, the fans are trying to keep lower speed. I think this is not what i want on my fans, as, if the temp goes in the 85-90 range, i want my fans going full speed. So, after the linear section, i would keep an exponential curve for the high temp section.
What do you think?

**Reply by ExFallenSkyTear** [reply]
I thought the same: the top part seems to suggest that the fans are kinda giving up on cooling, but in teory the the impact that a delta speed of 10% is influenced by the current fan speed. Basically, going from 0/10% to 10/20% has a much greater impact on temps than going 80 to 90, while the motor suffer much more at higher rpm; considering the noise too, I think there is no need to make an exponential curve.
Plus the higher the temperature, the higher the same rpm should cool it, because the cooling speed depends on the delta T, which is calculated by subtracting the enviromental temperature from the cpu temperature.
but I get your point, that's why I think that it would be better to give the user, the flexibility of entering a mathematical function from scratch.
I hope I made my self clear, cheers~

**Reply by Simone100190** [reply]
You got a point. As an engineer i understand your way of thinking. I am not into fluidodynamics modeling (and is a very complex topic too) so i dont have data on hand to study. I am debated between your explanation and "i must cool at fastest rate i can when temp goes too high". I can also point out that at higher temps actually the air get thinner, so we have: motor should being stressed less for putting out the same rpm, BUT at the same time less air is pulled through. 
I think, for now, is a matter of preferece: i do in fact prefer going all in with the RPM and maybe change the fan if the motor breaks, having all the possible cooling if the T skyrockets.

I think also that we need some kind of "Cooling Simulator" for developing the best curve for every set up 😄

**Reply by ExFallenSkyTear** [reply]
I'd love to know mathematically which curve is the best on term of efficiency, cooling, noise, etc., maybe a little overkill, but yeah.
anyway, back on the main topic, with a simple quadratic bezier would be easy to set up a EaseIn-like only type of curve, and I would use it too, probably.

**Reply by Simone100190** [reply]
Agree, more way to draw the curve the better!

**Comment by Jonathanese** [comment]
I thought I had already submitted something like this. But YES, THIS.

I'm so used to adjustable 3-point curves in FL Studio that I miss it in other software.

For those who don't use DAWs often, FL Studio's curves are customizable 3-point curves. So you have a start point, and endpoint, and a "tension handle". It isn't a Bezier specifically. But you can select what type of curve it is. "Single" curve which seems to be quadratic. "Double" is an S shape with the tension handle setting the slope, stair step with the tension handle setting the number of steps, and then others that aren't useful for fans, like oscillations and set-and-hold.

What Rem0o has is already way better than what SpeedFan ever had. But honestly if I had an adjustable curve option, I'd say basically all of my fan curves would be a 2 point single curve.

**Comment by ChronSyn** [comment]
I know this discussion hasn't received any responses for ~10 months, but thought I'd add my thoughts into the mix.

As others said, FanControl is a modern version of Speedfan but with a significantly better UI and more user-friendly features (owing to the advancements in software development over the years). It's a fantastic piece of software that allowed me to eliminate the need for iCUE while also negating the need for me to go into UEFI everytime I want to change fan curves (running on a Gigabyte X670 board, so no official Windows-based solution available from them). It's a fantastic piece of software and I feel like this is what other competing 'official' options should aspire towards.

With that in mind, I'm not necessarily looking for the entire graph to be bezier, but individual point control may be more effective to implement. For example, I might want a base-value to be linear to avoid it curving below a defined value. Then, I might want it to be bezier to the point after that. Then, I might want a linear stretch to the next point. etc.

With all that said, the control values are stored in the config JSON file (`Points` is an array of values, where each entry in the array is a string in the format `temperature, percentage`). It may be possible to use an external tool to generate curves and then have a way to import them. It's possible that such a tool doesn't exist already and one could be created. Taking this route is potentially much easier as it only requires implementing an 'add fan curve values' option (could be as simple as a text box and a confirmation button) and it also makes it one less tool for the author to maintain.
